src/tokens/mod.rs

```diff
@@ -0,0 +1,74 @@
+use self::token_iterator::Iter;
+use crate::{error::FluxError, matchers::MatcherName};
+use std::{fmt::Debug, ops::Range, rc::Rc};
+
+pub mod token_iterator;
+
+pub struct Token {
+    pub matcher_name: MatcherName,
+    pub matcher_id: usize,
+    pub children: Vec<Token>,
+    pub source: Rc<Vec<char>>,
+    pub range: Range<usize>,
+    pub failure: Option<FluxError>,
+}
+
+impl Token {
+    /// get the content the token is matching from the source
+    pub fn get_match(&self) -> String {
+        self.source[self.range.clone()].iter().collect()
+    }
+
+    /// return the name of the matcher that created the token
+    pub fn get_name(&self) -> &Option<String> {
+        &self.matcher_name
+    }
+
+    /// get the first child of the token
+    pub fn first(&self) -> Option<&Token> {
+        self.children.get(0)
+    }
+
+    /// get an iterator over all children of token
+    /// equivalent to calling `self.iter()`
+    pub fn all_children(&self) -> Iter {
+        Iter::new(self)
+    }
+
+    /// get an iterator over all children of `self` with a given `name`
+    pub fn named_children(&self, name: impl ToString) -> impl Iterator<Item = &Token> {
+        Iter::new(self)
+            .filter(move |t| matches!(t.matcher_name.as_ref(), Some(n) if n == &name.to_string()))
```

Comment:
Take in a reference to a string, or call to_string() once and hold it in a variable. Right now you're calling `name.to_string()` for every child, which is going to copy it for every comparison.

Reply:
what

Reply:
wait what

Reply:
I dont know why that was like that I changed it back to the way it originally was

Reply:
actually I do know why, I wanted to get rid of the extra line. I don't know if it does call `to_string` every time I think it might just call it once before the closure gets called